src/ripple/overlay/Slot.h

```diff
@@ -0,0 +1,635 @@
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_OVERLAY_SLOT_H_INCLUDED
+#define RIPPLE_OVERLAY_SLOT_H_INCLUDED
+
+#include <ripple/overlay/Peer.h>
+#include <ripple/overlay/Squelch.h>
+#include <ripple/overlay/SquelchCommon.h>
+#include <ripple/protocol/PublicKey.h>
+#include <ripple.pb.h>
+
+namespace ripple {
+
+namespace squelch {
+
+template <typename clock_type>
+class Slots;
+
+/** Peer's State */
+enum class PeerState : uint8_t {
+    Counting,   // counting messages
+    Selected,   // selected to relay, counting if Slot in Counting
+    Squelched,  // squelched, doesn't relay
+};
+/** Slot's State */
+enum class SlotState : uint8_t {
+    Counting,  // counting messages
+    Selected,  // peers selected, stop counting
+};
+
+template <typename Unit, typename TP>
+Unit
+epoch(TP const& t)
+{
+    return duration_cast<Unit>(t.time_since_epoch());
+}
+
+/** Abstract class. Declares squelch and unsquelch handlers */
+class SquelchHandler
+{
+public:
+    virtual ~SquelchHandler()
+    {
+    }
+    /** Squelch handler
+     * @param validator Public key of the source validator
+     * @param peer Peer to squelch
+     * @param duration Squelch duration in seconds
+     */
+    virtual void
+    squelch(
+        PublicKey const& validator,
+        std::weak_ptr<Peer> const& peer,
+        std::uint32_t duration) const = 0;
+    /** Unsquelch handler
+     * @param validator Public key of the source validator
+     * @param peer Peer to unsquelch
+     */
+    virtual void
+    unsquelch(PublicKey const& validator, std::weak_ptr<Peer> const& peer)
+        const = 0;
+};
+
+/**
+ * Slot is associated with a specific validator via validator's public key.
+ * Slot counts messages from a validator, selects peers to be the source
+ * of the messages, and communicates the peers to be squelched. Slot can be
+ * in the following states: 1) Counting. This is the peer selection state
+ * when Slot counts the messages and selects the peers; 2) Selected. Slot
+ * doesn't count messages in Selected state. A message received from
+ * unsquelched, disconnected peer, or idling peer may transition Slot to
+ * Counting state.
+ */
+template <typename clock_type>
```

Comment:
Now that this class is simplified to take a single template parameter, consider moving the implementation to a `.cpp` file and explicitly instantiating the clocks we need for linking.

Reply:
We'd have to move the ManualClock to a common header file, and we normally would not include a test file like that in the main code (tho I think we have a good reason in this case). But let's not worry about that now, we can do that later in another PR if we want.